src/lib/storage/table.cpp

```diff
@@ -106,8 +111,59 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
 }
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  // Typedef to limit word vomit
+  using abstract_ptr = std::shared_ptr<AbstractSegment>;
+
+  const auto compress_segment = [](const abstract_ptr& segment) -> abstract_ptr {
+#define TRY_COMPRESS_TYPE(Type, Segment)                        \
+  if (std::dynamic_pointer_cast<ValueSegment<Type>>(Segment)) { \
+    return std::make_shared<DictionarySegment<Type>>(Segment);  \
+  }
+    // Explicitly try to convert for each data type using macros for better readability
+    TRY_COMPRESS_TYPE(int32_t, segment);
+    TRY_COMPRESS_TYPE(int64_t, segment);
+    TRY_COMPRESS_TYPE(std::string, segment);
+    TRY_COMPRESS_TYPE(float, segment);
+    TRY_COMPRESS_TYPE(double, segment);
+
+#undef TRY_COMPRESS_TYPE
+
+    Fail("Invalid type for compression.");
+  };
+
+  Assert(chunk_id < chunk_count(), "Cannot compress chunk with Id " + std::to_string(chunk_id) + " out of bounce.");
+
+  // Immediately create new chunk to stop modification of the currently compressing chunk.
+  create_new_chunk();
+
+  // Keep currently compressing chunk for reading.
+  const auto chunk = get_chunk(chunk_id);
+  const auto segment_count = chunk->column_count();
+  auto futures = std::vector<std::future<abstract_ptr>>{};
+  futures.reserve(segment_count);
+
+  for (auto index = ColumnID{0}; index < segment_count; ++index) {
+    const auto segment = chunk->get_segment(index);
+    auto task = std::packaged_task<abstract_ptr(abstract_ptr)>(compress_segment);
+    auto future = task.get_future();
+    auto thread = std::thread(std::move(task), segment);
+
+    // Allow execution to continue in the background.
+    futures.push_back(std::move(future));
+    thread.detach();
+  }
+
+  // Create new empty chunk and append segments to it.
+  auto compressed_chunk = std::make_shared<Chunk>();
+  for (auto& future : futures) {
+    future.wait();
+    const auto compressed_segment = future.get();
+    compressed_chunk->add_segment(compressed_segment);
+  }
```

Comment:
TIL what futures are 😁 
Tbh, we implemented it differently, using threads and passing a function to each as well as the needed parameters and joining the threads in the end. But your way seems valid, in the end the performance evaluation by the teaching team will show which is faster 😎 

According to ChatGPT the reason for using futures is to not block the main thread. Was this also your reason to implement it this way? Do you know that there will a performance improvement because of this?